components/dashboard/src/start/CreateWorkspace.tsx

```diff
@@ -40,291 +42,290 @@ export interface CreateWorkspaceState {
-            this.setState({ error, stillParsing: false });
-        }
-    }
+export function CreateWorkspace({ contextUrl }: CreateWorkspaceProps) {
+    const [result, setResult] = useState<WorkspaceCreationResult>();
+    const [error, setError] = useState<StartWorkspaceError>();
+    const [selectAccountError, setSelectAccountError] = useState<SelectAccountPayload>();
+    const [stillParsing, setStillParsing] = useState(true);
+    const location = useLocation();
+    const org = useCurrentTeam();
+    const user = useCurrentUser();
+
+    const createWorkspace = useCallback(
+        (options?: Omit<GitpodServer.CreateWorkspaceOptions, "contextUrl">) => {
+            // Invalidate any previous result.
+            setResult(undefined);
+            setStillParsing(true);
+            setError(undefined);
+            // add options from search params
+            const opts = options || {};
+            Object.assign(opts, StartWorkspaceOptions.parseSearchParams(location.search));
+
+            (async () => {
+                // We assume anything longer than 3 seconds is no longer just parsing the context URL (i.e. it's now creating a workspace).
+                let timeout = setTimeout(() => setStillParsing(false), 3000);
+                try {
+                    const result = await getGitpodService().server.createWorkspace({
+                        contextUrl,
+                        // TODO (se): always pass the org id, once we've migrated all users to team-only attribution
+                        organizationId: !!user?.additionalData?.isMigratedToTeamOnlyAttribution ? org?.id : undefined,
+                        ...opts,
+                    });
+                    if (result.workspaceURL) {
+                        window.location.href = result.workspaceURL;
+                        return;
+                    }
+                    clearTimeout(timeout);
+                    setStillParsing(false);
+                    setResult(result);
+                } catch (error) {
+                    clearTimeout(timeout);
+                    console.error(error);
+                    setStillParsing(false);
+                    setError(error);
+                }
+            })();
+        },
+        [contextUrl, org, user, location.search],
+    );
 
-    async tryAuthorize(host: string, scopes?: string[]) {
-        try {
-            await openAuthorizeWindow({
-                host,
-                scopes,
-                onSuccess: () => {
-                    window.location.reload();
-                },
-                onError: (error) => {
-                    if (typeof error === "string") {
-                        try {
-                            const payload = JSON.parse(error);
-                            if (SelectAccountPayload.is(payload)) {
-                                this.setState({ selectAccountError: payload });
+    useEffect(() => {
+        createWorkspace();
+    }, [createWorkspace]);
```

Comment:
Trying to understand that:
* an effect bound to a callback 
  * when callback diffs call the callback

What's the intention here? 🤔 Haven't seen that pattern before.

Reply:
This should run on load and whenever the function has changed. That function changes when one of it's dependencies has changed.

Reply:
We don't need to debounce calls of `createWorkspace`?

Reply:
good point. I didn't experience that but it seems safer to guard this -> https://github.com/gitpod-io/gitpod/pull/16393